internal/display/bufwindow.go

```diff
@@ -405,8 +405,8 @@ func (w *BufWindow) displayBuffer() {
 				if r == bp[0] || r == bp[1] || rl == bp[0] || rl == bp[1] {
 					mb, left, found := b.FindMatchingBrace(bp, curLoc)
 					if found {
-						matchingBraces = append(matchingBraces, mb)
 						if !left {
+							matchingBraces = append(matchingBraces, mb)
```

Comment:
Thanks for noticing this.

Reply:
Hmm, no, this isn't ok: if `left` is true, it disables highlighting of the matching brace also in the case when braces are *not* next to each other. (And the same problem with underlining as well.)

Looks like we need a bit more complex logic here (a special check if braces are next to each other).

Reply:
I managed to solve that and most probably only the logic needs some tuning, **but** while testing it I came to the conclusion that the highlighting of matching braces left of the cursor is somehow a weird thing, because it handles the use case `((xyz))` totally wrong.

Just for the sake of completeness (potential to optimize, but I don't waste time with that, because the result doesn't fit my expectation):
```
				if r == bp[0] || r == bp[1] || rl == bp[0] || rl == bp[1] {
					mb, left, found := b.FindMatchingBrace(bp, curLoc)
					if found {
						if !left {
							matchingBraces = append(matchingBraces, mb)
							if b.Settings["matchbracestyle"].(string) != "highlight" {
								matchingBraces = append(matchingBraces, curLoc)
							}
						} else {
							if rl != bp[0] || r != bp[1] {
								matchingBraces = append(matchingBraces, mb)
							} else {
								if b.Settings["matchbracestyle"].(string) != "highlight" {
									matchingBraces = append(matchingBraces, curLoc)
								}
							}
							matchingBraces = append(matchingBraces, curLoc.Move(-1, b))
						}
					}
				}
```

I suggest to remove it, because from my perspective it's more what it should look like and e.g. `vim` doesn't do that either and so I will try to find a solution which fits to my expectation.

Reply:
Here you can find my proposal based on your last commit...

0001-Match-braces-only-in-case-the-curos-is-over-the-brac.patch:
```
From 9aa4e06e96bee4c2b8d0d4281df988dbc86b1f7c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?J=C3=B6ran=20Karl?=
 <3951388+JoeKar@users.noreply.github.com>
Date: Sat, 18 Nov 2023 21:01:42 +0100
Subject: [PATCH] Match braces only in case the curos is over the brace

...and ignore braces left of the cursor. This makes brace handling much easier
and less confusing.
---
 internal/action/actions.go    | 11 +++--------
 internal/buffer/buffer.go     | 37 +++++++++++------------------------
 internal/display/bufwindow.go | 15 +++++---------
 3 files changed, 19 insertions(+), 44 deletions(-)

diff --git a/internal/action/actions.go b/internal/action/actions.go
index 74e701f4..9fafebc0 100644
--- a/internal/action/actions.go
+++ b/internal/action/actions.go
@@ -1295,15 +1295,10 @@ func (h *BufPane) paste(clip string) {
 func (h *BufPane) JumpToMatchingBrace() bool {
 	for _, bp := range buffer.BracePairs {
 		r := h.Cursor.RuneUnder(h.Cursor.X)
-		rl := h.Cursor.RuneUnder(h.Cursor.X - 1)
-		if r == bp[0] || r == bp[1] || rl == bp[0] || rl == bp[1] {
-			matchingBrace, left, found := h.Buf.FindMatchingBrace(bp, h.Cursor.Loc)
+		if r == bp[0] || r == bp[1] {
+			matchingBrace, found := h.Buf.FindMatchingBrace(bp, h.Cursor.Loc)
 			if found {
-				if left {
-					h.Cursor.GotoLoc(matchingBrace)
-				} else {
-					h.Cursor.GotoLoc(matchingBrace.Move(1, h.Buf))
-				}
+				h.Cursor.GotoLoc(matchingBrace.Move(1, h.Buf))
 				h.Relocate()
 				return true
 			}
diff --git a/internal/buffer/buffer.go b/internal/buffer/buffer.go
index dc4d037f..606dc93f 100644
--- a/internal/buffer/buffer.go
+++ b/internal/buffer/buffer.go
@@ -995,28 +995,21 @@ var BracePairs = [][2]rune{
 // '{' and '}') as well as the location to match from
 // TODO: maybe can be more efficient with utf8 package
 // returns the location of the matching brace
-// if the boolean returned is true then the original matching brace is one character left
-// of the starting location
-func (b *Buffer) FindMatchingBrace(braceType [2]rune, start Loc) (Loc, bool, bool) {
+// if the boolean returned is true then the match was found
+func (b *Buffer) FindMatchingBrace(braceType [2]rune, start Loc) (Loc, bool) {
 	curLine := []rune(string(b.LineBytes(start.Y)))
 	startChar := ' '
 	if start.X >= 0 && start.X < len(curLine) {
 		startChar = curLine[start.X]
 	}
-	leftChar := ' '
-	if start.X-1 >= 0 && start.X-1 < len(curLine) {
-		leftChar = curLine[start.X-1]
-	}
 	var i int
-	if startChar == braceType[0] || leftChar == braceType[0] {
+	if startChar == braceType[0] {
 		for y := start.Y; y < b.LinesNum(); y++ {
 			l := []rune(string(b.LineBytes(y)))
 			xInit := 0
 			if y == start.Y {
 				if startChar == braceType[0] {
 					xInit = start.X
-				} else {
-					xInit = start.X - 1
 				}
 			}
 			for x := xInit; x < len(l); x++ {
@@ -1026,42 +1019,34 @@ func (b *Buffer) FindMatchingBrace(braceType [2]rune, start Loc) (Loc, bool, boo
 				} else if r == braceType[1] {
 					i--
 					if i == 0 {
-						if startChar == braceType[0] {
-							return Loc{x, y}, false, true
-						}
-						return Loc{x, y}, true, true
+						return Loc{x, y}, true
 					}
 				}
 			}
 		}
-	} else if startChar == braceType[1] || leftChar == braceType[1] {
+	} else if startChar == braceType[1] {
 		for y := start.Y; y >= 0; y-- {
 			l := []rune(string(b.lines[y].data))
 			xInit := len(l) - 1
 			if y == start.Y {
-				if leftChar == braceType[1] {
-					xInit = start.X - 1
-				} else {
+				if startChar == braceType[1] {
 					xInit = start.X
 				}
 			}
 			for x := xInit; x >= 0; x-- {
 				r := l[x]
-				if r == braceType[0] {
+				if r == braceType[1] {
+					i++
+				} else if r == braceType[0] {
 					i--
 					if i == 0 {
-						if leftChar == braceType[1] {
-							return Loc{x, y}, true, true
-						}
-						return Loc{x, y}, false, true
+						return Loc{x, y}, true
 					}
-				} else if r == braceType[1] {
-					i++
 				}
 			}
 		}
 	}
-	return start, true, false
+	return start, false
 }
 
 // Retab changes all tabs to spaces or vice versa
diff --git a/internal/display/bufwindow.go b/internal/display/bufwindow.go
index ae4525f6..13368082 100644
--- a/internal/display/bufwindow.go
+++ b/internal/display/bufwindow.go
@@ -401,17 +401,12 @@ func (w *BufWindow) displayBuffer() {
 				curLoc := c.Loc
 
 				r := c.RuneUnder(curX)
-				rl := c.RuneUnder(curX - 1)
-				if r == bp[0] || r == bp[1] || rl == bp[0] || rl == bp[1] {
-					mb, left, found := b.FindMatchingBrace(bp, curLoc)
+				if r == bp[0] || r == bp[1] {
+					mb, found := b.FindMatchingBrace(bp, curLoc)
 					if found {
-						if !left {
-							matchingBraces = append(matchingBraces, mb)
-							if b.Settings["matchbracestyle"].(string) != "highlight" {
-								matchingBraces = append(matchingBraces, curLoc)
-							}
-						} else {
-							matchingBraces = append(matchingBraces, curLoc.Move(-1, b))
+						matchingBraces = append(matchingBraces, mb)
+						if b.Settings["matchbracestyle"].(string) != "highlight" {
+							matchingBraces = append(matchingBraces, curLoc)
 						}
 					}
 				}
-- 
2.42.0


```

Use, if you want, but at least we should correct `FindMatchingBrace()` as suggested within this patch, because it does (at least from my perspective) the iteration to find the starting brace wrong.

Reply:
Yeah, I agree that this "match brace on the left of the cursor" feature is rather weird, and removing it would make the logic simpler and the behavior more intuitive and unambiguous.

I would perfectly live without this "feature" and I've been curious what was the original motivation for it, so I've dug a bit into its history.

1. Originally it was introduced ages ago in this PR: #1054 but as a separate `matchbraceleft` option, disabled by default.
2. Later, among other things, there was a fix for the issue #1101.
3. Then looks like this feature was lost during the big refactoring of micro.
4. Then it was reintroduced again in commit d1e713ce08ba15a59b62207109b4d6bca46c7a9f, but this time not as a separate option but combined with the default behavior (brace under the cursor).

As I understand, combining both behaviors has the advantage of having them both available at the same time (if anyone finds both behaviors useful). OTOH, it causes confusing ambiguities in cases like `((xyz))` or `(xyz)(xyz)`.

Perhaps we still can have both behavior combined, but should establish the rule: if there is a brace both under the cursor and on the left of it, the brace under the cursor always takes precedence. Looks like this is how already it already works for the `(xyz)(xyz)` case, so looks like we just need to make it work this way for the `((xyz))` case as well.

...Just noticed that the original author in #1054 mentioned that this feature was modeled after gedit. So I've just run gedit and I see it actually implements exactly what I've suggested above: it provides both "match brace under" and "match brace on the left" behaviors at the same time, and in ambiguous cases the "match brace under" behavior takes precedence.

Reply:
The documentation (of `matchbrace`: underline matching braces for '()', '{}', '[]' when the cursor is on a brace character.) never mentioned that it will fiddle around with the braces left of the cursor...it needs to be added there as well.

Furthermore, in case we care about the braces left of the cursor why we don't do for the braces right of the cursor?
For me this is still a bit inconsistent from look and feel. It was just a copy of a "feature" implemented in a different editor, but this doesn't mean that it has been a good "feature" :smile:.

Anyway, in case it should stay within micro then it definitely needs a fix, otherwise it can be considered as broken.

Reply:
> The documentation (of matchbrace: underline matching braces for '()', '{}', '[]' when the cursor is on a brace character.) never mentioned that it will fiddle around with the braces left of the cursor...it needs to be added there as well.

So we could change it to "...when the cursor is on a brace character or next to it" and be done.

> Furthermore, in case we care about the braces left of the cursor why we don't do for the braces right of the cursor?

I suppose it's because that would only serve aesthetic completeness, not practical usefulness.

I.e. I can imagine, for example, that someone is pressing `End` to move the cursor to the end of a line ending with an opening `{` brace, and is expecting to immediately see the matching closing `}` highlighted. The cursor will be not on the `{` but next to it, and this is a case when arguably the "match brace on the left" behavior may be useful. Are there any such real use cases for "match brace on the right"?

BTW speaking of gedit, it is a GUI text editor, not a terminal one, and I see it displays the cursor as a thin vertical line, not as a thick bar like terminals usually do. So it gedit it might seem to the user that the cursor is "between" characters, not on a character. I guess this justifies gedit's behavior with matching brace on the left but not on the right.

Reply:
> So we could change it to "...when the cursor is on a brace character or next to it" and be done.

:+1: 
@toiletbril
Do you take care of applying the last two patches and modifying the documentation accordingly?

@dmaluka 
Any complains regarding the last two patches?

> So it gedit it might seem to the user that the cursor is "between" characters, not on a character.

Ok, that's a quite good argument. Speaking of that clarifies, that in my imagined scenario (brace right of the cursor) the cursor is between the character before the brace and the character before that character.

And thus...

> Are there any such real use cases for "match brace on the right"?

...there is nothing more to add. :wink:

Reply:
> @dmaluka
Any complains regarding the last two patches?

Your 1st patch looks good to me, but the 2nd one - not quite... IMO this new logic in `displayBuffer()` is unclear (at least I'm too dumb to understand it) and is actually just a work around the fact that `FindMatchingBrace()` *wrongly* returns left=true.

Looks like we should instead just add such change in addition to your first patch:

```
diff --git a/internal/buffer/buffer.go b/internal/buffer/buffer.go
index d7376a22..424586d2 100644
--- a/internal/buffer/buffer.go
+++ b/internal/buffer/buffer.go
@@ -1008,7 +1008,7 @@ func (b *Buffer) FindMatchingBrace(braceType [2]rune, start Loc) (Loc, bool, boo
 		leftChar = curLine[start.X-1]
 	}
 	var i int
-	if startChar == braceType[0] || leftChar == braceType[0] {
+	if startChar == braceType[0] || (leftChar == braceType[0] && startChar != braceType[1]) {
 		for y := start.Y; y < b.LinesNum(); y++ {
 			l := []rune(string(b.LineBytes(y)))
 			xInit := 0
```

to make `FindMatchingBrace()` really always prefer left=false if the cursor is on a brace. And then we can restore the simple logic in `displayBuffer()` which was there before the last @toiletbril 's patch:

```
					mb, left, found := b.FindMatchingBrace(bp, curLoc)
					if found {
						matchingBraces = append(matchingBraces, mb)
						if !left {
							if b.Settings["matchbracestyle"].(string) != "highlight" {
								matchingBraces = append(matchingBraces, curLoc)
							}
						} else {
							matchingBraces = append(matchingBraces, curLoc.Move(-1, b))
						}
					}
```

Reply:
Yep, your solution is much cleaner and straight. Thanks to @dmaluka we can shrink the patch not to:
```
diff --git a/internal/buffer/buffer.go b/internal/buffer/buffer.go
index dc4d037f..424586d2 100644
--- a/internal/buffer/buffer.go
+++ b/internal/buffer/buffer.go
@@ -1008,7 +1008,7 @@ func (b *Buffer) FindMatchingBrace(braceType [2]rune, start Loc) (Loc, bool, boo
 		leftChar = curLine[start.X-1]
 	}
 	var i int
-	if startChar == braceType[0] || leftChar == braceType[0] {
+	if startChar == braceType[0] || (leftChar == braceType[0] && startChar != braceType[1]) {
 		for y := start.Y; y < b.LinesNum(); y++ {
 			l := []rune(string(b.LineBytes(y)))
 			xInit := 0
@@ -1039,24 +1039,24 @@ func (b *Buffer) FindMatchingBrace(braceType [2]rune, start Loc) (Loc, bool, boo
 			l := []rune(string(b.lines[y].data))
 			xInit := len(l) - 1
 			if y == start.Y {
-				if leftChar == braceType[1] {
-					xInit = start.X - 1
-				} else {
+				if startChar == braceType[1] {
 					xInit = start.X
+				} else {
+					xInit = start.X - 1
 				}
 			}
 			for x := xInit; x >= 0; x-- {
 				r := l[x]
-				if r == braceType[0] {
+				if r == braceType[1] {
+					i++
+				} else if r == braceType[0] {
 					i--
 					if i == 0 {
-						if leftChar == braceType[1] {
-							return Loc{x, y}, true, true
+						if startChar == braceType[1] {
+							return Loc{x, y}, false, true
 						}
-						return Loc{x, y}, false, true
+						return Loc{x, y}, true, true
 					}
-				} else if r == braceType[1] {
-					i++
 				}
 			}
 		}
diff --git a/internal/display/bufwindow.go b/internal/display/bufwindow.go
index ae4525f6..bca6bf8e 100644
--- a/internal/display/bufwindow.go
+++ b/internal/display/bufwindow.go
@@ -405,8 +405,8 @@ func (w *BufWindow) displayBuffer() {
 				if r == bp[0] || r == bp[1] || rl == bp[0] || rl == bp[1] {
 					mb, left, found := b.FindMatchingBrace(bp, curLoc)
 					if found {
+						matchingBraces = append(matchingBraces, mb)
 						if !left {
-							matchingBraces = append(matchingBraces, mb)
 							if b.Settings["matchbracestyle"].(string) != "highlight" {
 								matchingBraces = append(matchingBraces, curLoc)
 							}
diff --git a/runtime/help/options.md b/runtime/help/options.md
index 4ea65e8d..7edc97ef 100644
--- a/runtime/help/options.md
+++ b/runtime/help/options.md
@@ -211,7 +211,7 @@ Here are the available options:
 	default value: `false`
 
 * `matchbrace`: show matching braces for '()', '{}', '[]' when the cursor
-   is on a brace character.
+   is on a brace character or next to it.
 
     default value: `true`
 

```

Thanks for co-authoring, reviewing and hopefully getting it done/merged.

Reply:
You guys went all out on this pull request. Thanks for help!